### PR TITLE
ioda-converter for SMOS SSS L2 from ESA

### LINKED ISSUE #489: ioda-converter for SSS SMOS data from ESA
## Description
Ioda-converter in python exists for SSS data from SMAP RSS feed. 
A new converter is provided for SSS data from SMOS provided by ESA in NetCDF format and converted to IODA NetCDF data assimilation.

The data being converted is in L2.

### PR DESCRIPTION
**Pull request by ShastriPaturi**
## Description

This PR adds a ioda-converter for SSS SMOS L2 data from ESA (smos_sss2ioda.py)
The ioda-converter reads the date from each data (nc) file per day since there is no "time" variable provided.
The SMOS ESA data is available from 20100601

What problem does it fix? What new capability does it add?
Adds capability to use SMOS SSS data

[See the JEDI Documentation](https://jointcenterforsatellitedataassimilation-jedi-docs.readthedocs-hosted.com/en/latest/inside/practices/pullrequest.html) for further information.

### Issue(s) addressed

Link the issues to be closed with this PR
- fixes #489 >

### CODE REVIEW

**Comment by ShastriPaturi** [comment]
@HamidehGMAO and @guillaumevernieres, why is it showing 3 failing checks here?

**Comment by HamidehGMAO** [comment]
I can't login in aws right now, @srherbener could you please see why AWS tests are failing here? Thanks

**Comment by HamidehGMAO** [comment]
@ShastriPaturi could you please add a sample file from SMOS here https://github.com/JCSDA-internal/ioda-converters/tree/develop/test/testinput, so I could test this converter with that. Thanks

**Comment by ShastriPaturi** [comment]
> @ShastriPaturi could you please add a sample file from SMOS here https://github.com/JCSDA-internal/ioda-converters/tree/develop/test/testinput, so I could test this converter with that. Thanks

@HamidehGMAO I added it to my feature branch

**Comment by srherbener** [comment]
> I can't login in aws right now, @srherbener could you please see why AWS tests are failing here? Thanks

I also cannot login. I'll try to find out what is going on.

**Comment by HamidehGMAO** [comment]
> > I can't login in aws right now, @srherbener could you please see why AWS tests are failing here? Thanks
> 
> I also cannot login. I'll try to find out what is going on.

Thanks @srherbener

**Comment by srherbener** [comment]
@HamidehGMAO try using "jcsds-noaa" for the account ID (instead of "jcsda") when logging into AWS.

**Comment by srherbener** [comment]
The test failures are coding norms errors:
```
test 5
--
909 | Start 5: iodaconv_marine_coding_norms
910 |  
911 | 5: Test command: /build_container/bin/iodaconv_lint.sh "/jcsda/ioda-bundle/iodaconv/src/marine" "/jcsda/ioda-bundle/iodaconv"
912 | 5: Environment variables:
913 | 5:  OMP_NUM_THREADS=1
914 | 5: Test timeout computed to be: 10000000
915 | 5: + SRC_DIR=/jcsda/ioda-bundle/iodaconv/src/marine
916 | 5: + PYCODESTYLE_CFG_PATH=/jcsda/ioda-bundle/iodaconv
917 | 5: + cd /jcsda/ioda-bundle/iodaconv/src/marine
918 | 5: + pycodestyle -v --config=/jcsda/ioda-bundle/iodaconv/.pycodestyle --filename=*.py,*.py.in .
919 | 5: cli configuration: /jcsda/ioda-bundle/iodaconv/.pycodestyle
920 | 5: directory .
921 | 5: checking ./argoClim2ioda.py
922 | 5: checking ./cryosat_ice2ioda.py
923 | 5: checking ./emc_ice2ioda.py
924 | 5: checking ./gds2_sst2ioda.py
925 | 5: checking ./glider2ioda.py
926 | 5: checking ./gmao_obs2ioda.py
927 | 5: checking ./godae_profile2ioda.py
928 | 5: checking ./godae_ship2ioda.py
929 | 5: checking ./godae_trak2ioda.py
930 | 5: checking ./hgodas_adt2ioda.py
931 | 5: checking ./hgodas_insitu2ioda.py
932 | 5: checking ./hgodas_sst2ioda.py
933 | 5: checking ./ndbc_hfradar2ioda.py
934 | 5: checking ./rads_adt2ioda.py
935 | 5: checking ./smap_sss2ioda.py
936 | 5: checking ./smos_sss2ioda.py
937 | 5: ./smos_sss2ioda.py:55:1: W293 blank line contains whitespace
938 | 5: ./smos_sss2ioda.py:57:36: E231 missing whitespace after ','
939 | 5: ./smos_sss2ioda.py:59:1: W293 blank line contains whitespace
940 | 5: ./smos_sss2ioda.py:61:18: E126 continuation line over-indented for hanging indent
941 | 5: ./smos_sss2ioda.py:64:27: E231 missing whitespace after ':'
942 | 5: ./smos_sss2ioda.py:66:17: E121 continuation line under-indented for hanging indent
943 | 5: ./smos_sss2ioda.py:67:16: E221 multiple spaces before operator
944 | 5: ./smos_sss2ioda.py:68:16: E221 multiple spaces before operator
945 | 5: ./smos_sss2ioda.py:69:16: E221 multiple spaces before operator
946 | 5: ./smos_sss2ioda.py:71:19: E221 multiple spaces before operator
947 | 5: ./smos_sss2ioda.py:83:20: E111 indentation is not a multiple of four
948 | 5: ./smos_sss2ioda.py:85:20: E111 indentation is not a multiple of four
949 | 5: ./smos_sss2ioda.py:86:17: E741 ambiguous variable name 'I'
950 | 5: ./smos_sss2ioda.py:86:18: E225 missing whitespace around operator
951 | 5: ./smos_sss2ioda.py:87:22: E225 missing whitespace around operator
952 | 5: ./smos_sss2ioda.py:88:20: E225 missing whitespace around operator
953 | 5: ./smos_sss2ioda.py:89:20: E225 missing whitespace around operator
954 | 5: ./smos_sss2ioda.py:90:20: E225 missing whitespace around operator
955 | 5: ./smos_sss2ioda.py:91:24: E225 missing whitespace around operator
956 | 5: ./smos_sss2ioda.py:91:26: E201 whitespace after '('
957 | 5: ./smos_sss2ioda.py:91:62: E231 missing whitespace after ','
958 | 5: ./smos_sss2ioda.py:91:79: W504 line break after binary operator
959 | 5: ./smos_sss2ioda.py:91:80: W291 trailing whitespace
960 | 5: ./smos_sss2ioda.py:92:50: E231 missing whitespace after ','
961 | 5: ./smos_sss2ioda.py:92:60: E202 whitespace before ')'
962 | 5: ./smos_sss2ioda.py:93:25: E225 missing whitespace around operator
963 | 5: ./smos_sss2ioda.py:93:49: E231 missing whitespace after ','
964 | 5: ./smos_sss2ioda.py:99:1: W293 blank line contains whitespace
965 | 5: ./smos_sss2ioda.py:101:13: E303 too many blank lines (2)
966 | 5: ./smos_sss2ioda.py:126:1: W293 blank line contains whitespace
967 | 5: ./smos_sss2ioda.py:139:1: E305 expected 2 blank lines after class or function definition, found 1
968 | 5: checking ./viirs_modis_l2_oc2ioda.py
969 | 5: checking ./viirs_modis_l3_oc2ioda.py
970 | 5: 2       E111 indentation is not a multiple of four
971 | 5: 1       E121 continuation line under-indented for hanging indent
972 | 5: 1       E126 continuation line over-indented for hanging indent
973 | 5: 1       E201 whitespace after '('
974 | 5: 1       E202 whitespace before ')'
975 | 5: 4       E221 multiple spaces before operator
976 | 5: 7       E225 missing whitespace around operator
977 | 5: 5       E231 missing whitespace after ','
978 | 5: 1       E303 too many blank lines (2)
979 | 5: 1       E305 expected 2 blank lines after class or function definition, found 1
980 | 5: 1       E741 ambiguous variable name 'I'
981 | 5: 1       W291 trailing whitespace
982 | 5: 4       W293 blank line contains whitespace
983 | 5: 1       W504 line break after binary operator
984 | 1/1 Test #5: iodaconv_marine_coding_norms .....***Failed    0.36 sec
985 |  
986 | 0% tests passed, 1 tests failed out of 1
987 |  
988 | Label Time Summary:
989 | iodaconv    =   0.36 sec*proc (1 test)
990 | script      =   0.36 sec*proc (1 test)
991 |  
992 | Total Test time (real) =   0.36 sec
993 |  
994 | The following tests FAILED:
995 | 5 - iodaconv_marine_coding_norms (Failed)
996 | Errors while running CTest
```

**Comment by HamidehGMAO** [comment]
> jcsds-noaa

OK, Thanks for your quick reply @srherbener

**Comment by ShastriPaturi** [comment]
@HamidehGMAO and @guillaumevernieres, I have tested the ctest for th SMOS converter and have only used the ioda-converter repo and not the ioda-bundle. The bufr and coding norms are failing.
I want to

**Comment by ShastriPaturi** [comment]
> @HamidehGMAO and @guillaumevernieres, I have tested the ctest for th SMOS converter and have only used the ioda-converter repo and not the ioda-bundle. The bufr and coding norms are failing.

**Review by srherbener — CHANGES_REQUESTED** [review]
I think the code looks good and it is great that you added a fast running unit test! The tests are failing for two reasons - coding norms errors and an entry for the new script is missing from the CMakeLists.txt configuration:

https://github.com/JCSDA-internal/ioda-converters/blob/e037c2572f17e1261e17c452f44902440596456a/src/marine/CMakeLists.txt#L15

Just copy and paste the smap_sss2ioda.py line to below that line and change it to smos_sss2ioda.py to fix the CMakeLists.txt configuration.

Thanks!

**Comment by ShastriPaturi** [comment]
> I think the code looks good and it is great that you added a fast running unit test! The tests are failing for two reasons - coding norms errors and an entry for the new script is missing from the CMakeLists.txt configuration:
> 
> https://github.com/JCSDA-internal/ioda-converters/blob/e037c2572f17e1261e17c452f44902440596456a/src/marine/CMakeLists.txt#L15
> 
> Just copy and paste the smap_sss2ioda.py line to below that line and change it to smos_sss2ioda.py to fix the CMakeLists.txt configuration.
> 
> Thanks!

I made the change you requested. I though I added. Forgot to commit.

File: src/marine/smos_sss2ioda.py [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+#
+
+from __future__ import print_function
+import sys
+import argparse
+import numpy as np
+from datetime import datetime, timedelta
+import netCDF4 as nc
+import re
+import dateutil.parser
```
Comment (aerorahul) [inline src/marine/smos_sss2ioda.py:17]: don't think these are used anywhere in the script.

File: src/marine/smos_sss2ioda.py [hunk 2]
```diff
@@ -0,0 +1,140 @@
+#!/usr/bin/env python
+
+#
+# (C) Copyright 2019 UCAR
+#
+# This software is licensed under the terms of the Apache Licence Version 2.0
+# which can be obtained at http://www.apache.org/licenses/LICENSE-2.0.
+#
+
+from __future__ import print_function
```
Comment (aerorahul) [inline src/marine/smos_sss2ioda.py:10]: don't need this.

**Comment by srherbener** [comment]
@ShastriPaturi thank you for your updates! The smos test is passing now and you are down to one test failure which is due to coding norms errors in the new smos_sss2ioda.py script.
```
test 5
--
909 | Start 5: iodaconv_marine_coding_norms
910 |  
911 | 5: Test command: /build_container/bin/iodaconv_lint.sh "/jcsda/ioda-bundle/iodaconv/src/marine" "/jcsda/ioda-bundle/iodaconv"
912 | 5: Environment variables:
913 | 5:  OMP_NUM_THREADS=1
914 | 5: Test timeout computed to be: 10000000
915 | 5: + SRC_DIR=/jcsda/ioda-bundle/iodaconv/src/marine
916 | 5: + PYCODESTYLE_CFG_PATH=/jcsda/ioda-bundle/iodaconv
917 | 5: + cd /jcsda/ioda-bundle/iodaconv/src/marine
918 | 5: + pycodestyle -v --config=/jcsda/ioda-bundle/iodaconv/.pycodestyle --filename=*.py,*.py.in .
919 | 5: cli configuration: /jcsda/ioda-bundle/iodaconv/.pycodestyle
920 | 5: directory .
921 | 5: checking ./argoClim2ioda.py
922 | 5: checking ./cryosat_ice2ioda.py
923 | 5: checking ./emc_ice2ioda.py
924 | 5: checking ./gds2_sst2ioda.py
925 | 5: checking ./glider2ioda.py
926 | 5: checking ./gmao_obs2ioda.py
927 | 5: checking ./godae_profile2ioda.py
928 | 5: checking ./godae_ship2ioda.py
929 | 5: checking ./godae_trak2ioda.py
930 | 5: checking ./hgodas_adt2ioda.py
931 | 5: checking ./hgodas_insitu2ioda.py
932 | 5: checking ./hgodas_sst2ioda.py
933 | 5: checking ./ndbc_hfradar2ioda.py
934 | 5: checking ./rads_adt2ioda.py
935 | 5: checking ./smap_sss2ioda.py
936 | 5: checking ./smos_sss2ioda.py
937 | 5: ./smos_sss2ioda.py:55:1: W293 blank line contains whitespace
938 | 5: ./smos_sss2ioda.py:57:36: E231 missing whitespace after ','
939 | 5: ./smos_sss2ioda.py:59:1: W293 blank line contains whitespace
940 | 5: ./smos_sss2ioda.py:61:18: E126 continuation line over-indented for hanging indent
941 | 5: ./smos_sss2ioda.py:64:27: E231 missing whitespace after ':'
942 | 5: ./smos_sss2ioda.py:66:17: E121 continuation line under-indented for hanging indent
943 | 5: ./smos_sss2ioda.py:67:16: E221 multiple spaces before operator
944 | 5: ./smos_sss2ioda.py:68:16: E221 multiple spaces before operator
945 | 5: ./smos_sss2ioda.py:69:16: E221 multiple spaces before operator
946 | 5: ./smos_sss2ioda.py:71:19: E221 multiple spaces before operator
947 | 5: ./smos_sss2ioda.py:83:20: E111 indentation is not a multiple of four
948 | 5: ./smos_sss2ioda.py:85:20: E111 indentation is not a multiple of four
949 | 5: ./smos_sss2ioda.py:86:17: E741 ambiguous variable name 'I'
950 | 5: ./smos_sss2ioda.py:86:18: E225 missing whitespace around operator
951 | 5: ./smos_sss2ioda.py:87:22: E225 missing whitespace around operator
952 | 5: ./smos_sss2ioda.py:88:20: E225 missing whitespace around operator
953 | 5: ./smos_sss2ioda.py:89:20: E225 missing whitespace around operator
954 | 5: ./smos_sss2ioda.py:90:20: E225 missing whitespace around operator
955 | 5: ./smos_sss2ioda.py:91:24: E225 missing whitespace around operator
956 | 5: ./smos_sss2ioda.py:91:26: E201 whitespace after '('
957 | 5: ./smos_sss2ioda.py:91:62: E231 missing whitespace after ','
958 | 5: ./smos_sss2ioda.py:91:79: W504 line break after binary operator
959 | 5: ./smos_sss2ioda.py:91:80: W291 trailing whitespace
960 | 5: ./smos_sss2ioda.py:92:50: E231 missing whitespace after ','
961 | 5: ./smos_sss2ioda.py:92:60: E202 whitespace before ')'
962 | 5: ./smos_sss2ioda.py:93:25: E225 missing whitespace around operator
963 | 5: ./smos_sss2ioda.py:93:49: E231 missing whitespace after ','
964 | 5: ./smos_sss2ioda.py:99:1: W293 blank line contains whitespace
965 | 5: ./smos_sss2ioda.py:101:13: E303 too many blank lines (2)
966 | 5: ./smos_sss2ioda.py:126:1: W293 blank line contains whitespace
967 | 5: ./smos_sss2ioda.py:139:1: E305 expected 2 blank lines after class or function definition, found 1
968 | 5: checking ./viirs_modis_l2_oc2ioda.py
969 | 5: checking ./viirs_modis_l3_oc2ioda.py
970 | 5: 2       E111 indentation is not a multiple of four
971 | 5: 1       E121 continuation line under-indented for hanging indent
972 | 5: 1       E126 continuation line over-indented for hanging indent
973 | 5: 1       E201 whitespace after '('
974 | 5: 1       E202 whitespace before ')'
975 | 5: 4       E221 multiple spaces before operator
976 | 5: 7       E225 missing whitespace around operator
977 | 5: 5       E231 missing whitespace after ','
978 | 5: 1       E303 too many blank lines (2)
979 | 5: 1       E305 expected 2 blank lines after class or function definition, found 1
980 | 5: 1       E741 ambiguous variable name 'I'
981 | 5: 1       W291 trailing whitespace
982 | 5: 4       W293 blank line contains whitespace
983 | 5: 1       W504 line break after binary operator
984 | 1/1 Test #5: iodaconv_marine_coding_norms .....***Failed    0.34 sec
985 |  
986 | 0% tests passed, 1 tests failed out of 1
987 |  
988 | Label Time Summary:
989 | iodaconv    =   0.34 sec*proc (1 test)
990 | script      =   0.34 sec*proc (1 test)
991 |  
992 | Total Test time (real) =   0.35 sec
993 |  
994 | The following tests FAILED:
995 | 5 - iodaconv_marine_coding_norms (Failed)
996 | Errors while running CTest
```

File: src/marine/smos_sss2ioda.py [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+            sss_err = sss_err[mask]
+            sss_qc = sss_qc[mask]
+
+            for i in range(len(lon)):
+                if sss_qc[i] <= 150:
+                   sss_qc[i] = 0
+                else:
+                   sss_qc[i] = 1
+                # get date from filename
+                I = f.find("SM_")
```
Comment (aerorahul) [inline src/marine/smos_sss2ioda.py:84]: not to pick on this, but you are using lowercase `i` and uppercase `I` in the same loop.

File: src/marine/smos_sss2ioda.py [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+                 'sss_qc': 'Dg_quality_SSS_corr',
+                }
+            lon = ncd.variables['Longitude'][:]
+            lat = ncd.variables['Latitude'][:]
+            sss = ncd.variables['SSS_corr'][:]
+            sss_err = ncd.variables['Sigma_SSS_corr'][:]
+            sss_qc = ncd.variables['Dg_quality_SSS_corr'][:]
+            sss_qc = sss_qc.astype(int)
+
+	    mask = np.logical_not(sss.mask)
```
Comment (aerorahul) [inline src/marine/smos_sss2ioda.py:71]: either this line or the next set of lines are indented out of place.

**Review by aerorahul — CHANGES_REQUESTED** [review]
see comments.

**Comment by ShastriPaturi** [comment]
@aerorahul and @srherbener, I made the changes as requested and coding_norms test passes.

**Comment by aerorahul** [comment]
@ShastriPaturi 
marine_coding_norms are failing.
```

Start  5: iodaconv_marine_coding_norms
--
761 | 5/63 Test  #5: iodaconv_marine_coding_norms .............***Failed    0.36 sec

```

**Comment by ShastriPaturi** [comment]
@aerorahul please check now.

image

**Review by aerorahul — APPROVED** [review]
looks good.

**Review by srherbener — APPROVED** [review]
thanks for addressing my comments!

**Comment by ShastriPaturi** [comment]
Thank you @srherbener and @aerorahul

**Pull request merged**